source/fab/database.py

```diff
@@ -104,27 +104,30 @@ class SqliteStateDatabase(StateDatabase):
     '''
     def __init__(self, working_directory: Path):
         self._working_directory: Path = working_directory
-
-        if not self._working_directory.exists():
-            self._working_directory.mkdir(parents=True)
-
-        self._connection: sqlite3.Connection \
-            = sqlite3.connect(str(working_directory / 'state.db'))
-        self._connection.row_factory = sqlite3.Row
+        self._connection: Optional[sqlite3.Connection] = None
 
     def __del__(self):
-        self._connection.close()
+        if self._connection is not None:
+            self._connection.close()
+
+    def get_connection(self) -> sqlite3.Connection:
```

Comment:
I don't think we should be spilling sqlite objects such as the connection out of this class. I know that Python doesn't actually support data hiding but we can at least pretend it does by preceding the method name with an underline.

Reply:
Agreed, I'm actually not sure why I didn't do that to begin with